File: app/src/main/java/com/krystianwsul/checkme/firebase/json/JsonWrapper.java
package com.krystianwsul.checkme.firebase.json;

import android.support.annotation.NonNull;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.google.firebase.database.IgnoreExtraProperties;

import junit.framework.Assert;

import java.util.Map;
import java.util.Set;

@SuppressWarnings({"unused", "WeakerAccess"})
@IgnoreExtraProperties
public class JsonWrapper {
    public Map<String, Boolean> recordOf;
    public ProjectJson projectJson;

    public JsonWrapper() {

    }

    public JsonWrapper(@NonNull Set<String> recordOf, @NonNull ProjectJson projectJson) {
        Assert.assertTrue(!recordOf.isEmpty());

        this.recordOf = Stream.of(recordOf)
                .collect(Collectors.toMap(friend -> friend, friend -> true));

        this.projectJson = projectJson;
    }

    public void updateRecordOf(@NonNull Set<String> add, @NonNull Set<String> remove) {
        Assert.assertTrue(Stream.of(add).noneMatch(remove::contains));
        Assert.assertTrue(Stream.of(add).noneMatch(recordOf::containsKey));

        Assert.assertTrue(Stream.of(remove).noneMatch(add::contains));
        Assert.assertTrue(Stream.of(remove).allMatch(recordOf::containsKey));

        for (String key : remove)
            recordOf.remove(key);

        for (String key : add)
            recordOf.put(key, true);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/json/JsonWrapper.kt
package com.krystianwsul.checkme.firebase.json


class JsonWrapper @JvmOverloads constructor(
        val recordOf: MutableMap<String, Boolean> = mutableMapOf(),
        var projectJson: SharedProjectJson = SharedProjectJson()) {

    constructor(recordOf: Set<String>, projectJson: SharedProjectJson) : this(recordOf.associateBy({ it }, { true }).toMutableMap(), projectJson)

    fun updateRecordOf(add: Set<String>, remove: Set<String>) {
        check(add.none { remove.contains(it) })
        check(add.none { recordOf.containsKey(it) })

        check(remove.none { add.contains(it) })
        check(remove.all { recordOf.containsKey(it) })

        for (key in remove)
            recordOf.remove(key)

        for (key in add)
            recordOf[key] = true
    }
}
